### PR TITLE
Re-architectured integration tests

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+    assertThat(result).hasSize(1);
+    List<Row> filteredResult = result.stream().filter(row -> row.getLong(0) == 42L)
+        .collect(Collectors.toList());
+    assertThat(filteredResult).hasSize(1);
+  }
+
+  @Test
+  public void testOutOfOrderColumns() {
+    Row row = spark.read().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:122]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        .option("table", SHAKESPEARE_TABLE)
+        .option("readDataFormat", dataFormat).load()
+        .select("word_count", "word").head();
+    assertThat(row.get(0)).isInstanceOf(Long.class);
+    assertThat(row.get(1)).isInstanceOf(String.class);
+  }
+
+  @Test
+  public void testSelectAllColumnsFromATable() {
+    Row row = spark.read().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:132]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        .select("word_count", "word", "corpus", "corpus_date").head();
+    assertThat(row.get(0)).isInstanceOf(Long.class);
+    assertThat(row.get(1)).isInstanceOf(String.class);
+    assertThat(row.get(2)).isInstanceOf(String.class);
+    assertThat(row.get(3)).isInstanceOf(Long.class);
+  }
+
+  @Test
+  public void testNumberOfPartitions() {
+    Dataset<Row> df = spark.read().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:144]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        .option("table", LARGE_TABLE)
+        .option("parallelism", "5")
+        .option("readDataFormat", dataFormat)
+        .load();
+    assertThat(df.rdd().getNumPartitions()).isEqualTo(5);
+  }
+
+  @Test
+  public void testDefaultNumberOfPartitions() {
+    Dataset<Row> df = spark.read().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:154]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        .load();
+
+    assertThat(df.rdd().getNumPartitions()).isEqualTo(58);
+  }
+
+  @Test(timeout = 300_000)
+  public void testBalancedPartitions() {
+    // Select first partition
+    Dataset<Row> df = spark.read()
+        .format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:166]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    long numRowsLowerBound = LARGE_TABLE_NUM_ROWS / df.rdd().getNumPartitions();
+    assertThat(numRowsLowerBound <= sizeOfFirstPartition).isTrue();
+    assertThat(sizeOfFirstPartition < numRowsLowerBound * 1.1).isTrue();
+  }
+
+  @Test
+  public void testKeepingFiltersBehaviour() {
+    Set<String> newBehaviourWords = extractWords(
+        spark.read().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:192]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+  @Test
+  public void testColumnOrderOfStruct() {
+    StructType schema = Encoders.bean(TestConstants.ColumnOrderTestClass.class).schema();
+
+    Dataset<TestConstants.ColumnOrderTestClass> dataset = spark.read()
+        .schema(schema)
+        .option("dataset", testDataset.toString())
+        .option("table", STRUCT_COLUMN_ORDER_TEST_TABLE_NAME)
+        .format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:218]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.MetadataBuilder;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.junit.Ignore;
+import org.junit.Test;
+
+public class ReadIntegrationTestBase extends SparkBigQueryIntegrationTestBase {
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:42]: should'nt we run the tests in this file for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+    return bq.getTable(testDataset.toString(), testTable + "_partitioned")
+        .getDefinition();
+  }
+
+  protected void writeToBigQuery(Dataset<Row> df, SaveMode mode) {
+    writeToBigQuery(df, mode, "parquet");
+  }
+
+  protected void writeToBigQuery(Dataset<Row> df, SaveMode mode, String format) {
+    df.write().format("bigquery")
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java:134]: should'nt we test for V2 as well?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+          new NumStruct(
+              3L,
+              2L,
+              1L,
+              ImmutableList.of(
+                  new StringStruct("0:str3", "0:str1", "0:str2"),
+                  new StringStruct("1:str3", "1:str1", "1:str2"))),
+          "outer_string");
+
+  public static class StringStruct {
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:197]: Can we move it to `model` package similar to `Friend.java`?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+    public String toString() {
+      return "StringStruct{" +
+          "str3='" + str3 + '\'' +
+          ", str1='" + str1 + '\'' +
+          ", str2='" + str2 + '\'' +
+          '}';
+    }
+  }
+
+  public static class NumStruct {
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:262]: Can we move it to `model` package similar to `Friend.java`?

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+      return "NumStruct{" +
+          "num3=" + num3 +
+          ", num2=" + num2 +
+          ", num1=" + num1 +
+          ", stringStructArr=" + strings +
+          '}';
+    }
+  }
+
+  public static class ColumnOrderTestClass {
```
Comment (himanshukohli09) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:340]: Can we move it to `model` package similar to `Friend.java`?

**Review by himanshukohli09 — CHANGES_REQUESTED** [review]

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java (hunk 9 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java:134]: V2 will be "bigquery" as well, as it is on a separate jar

File: spark-bigquery-dsv1/src/test/java/com/google/cloud/spark/bigquery/integration/DataSourceV1ReadFromQueryIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package com.google.cloud.spark.bigquery.integration;
+
+import org.apache.spark.sql.SparkSession;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+public class DataSourceV1ReadFromQueryIntegrationTest extends ReadFromQueryIntegrationTestBase {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv1/src/test/java/com/google/cloud/spark/bigquery/integration/DataSourceV1ReadFromQueryIntegrationTest.java:7]: tests are from the super-class

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java (hunk 8 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:42]: Yes - in the last commit

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24ReadFromQueryIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,5 @@
+package com.google.cloud.spark.bigquery.integration;
+
+public class Spark24ReadFromQueryIntegrationTest extends ReadFromQueryIntegrationTestBase {
+
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24ReadFromQueryIntegrationTest.java:4]: tests come from the super class

File: spark-bigquery-dsv1/src/test/java/com/google/cloud/spark/bigquery/integration/DataSourceV1WriteIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.spark.sql.types.StructType;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import scala.collection.JavaConverters;
+import scala.collection.Seq;
+
+public class DataSourceV1WriteIntegrationTest extends WriteIntegrationTestBase {
+
+  // DSv2 does not support BigNumeric yet
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv1/src/test/java/com/google/cloud/spark/bigquery/integration/DataSourceV1WriteIntegrationTest.java:27]: Waiting for PR #449

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,5 @@
+package com.google.cloud.spark.bigquery.integration;
+
+public class Spark24WriteIntegrationTest extends WriteIntegrationTestBase {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java:3]: tests are in the superclass

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+  private static final String  ALL_TYPES_TABLE_NAME = "all_types";
+  private BigQuery bq;
+
+  protected ReadFromQueryIntegrationTestBase() {
+    super();
+    this.bq = BigQueryOptions.getDefaultInstance().getService();
+  }
+
+  private void testReadFromQueryInternal(String query) {
+    Dataset<Row> df = spark.read().format("bigquery")
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java:41]: It is in the common v1/v2 test code

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    assertThat(corpuses).containsExactlyElementsIn(expectedCorpuses);
+  }
+
+  @Test public void
+  testBadQuery() {
+    String badSql = "SELECT bogus_column FROM `bigquery-public-data.samples.shakespeare`";
+    // v1 throws BigQueryConnectorException
+    // v2 throws Guice ProviderException, as the table is materialized in teh module
+    assertThrows(RuntimeException.class, () -> {
+      spark.read().format("bigquery")
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java:106]: It is in the common v1/v2 test code

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/model/Data.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
+package com.google.cloud.spark.bigquery.integration.model;
+
+import com.google.common.base.Objects;
+import java.io.Serializable;
+import java.sql.Timestamp;
+
+public class Data implements Serializable {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/model/Data.java:7]: It is generic on purpose

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:132]: v2 format name is "bigquery"

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java (hunk 3 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:144]: v2 format name is "bigquery"

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java (hunk 5 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:166]: v2 format name is "bigquery"

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java (hunk 6 above)
Comment (davidrabinowitz) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadByFormatIntegrationTestBase.java:158]: In v2 we have only the new behavior

File: cloudbuild/cloudbuild.yaml [hunk 20]
```diff
@@ -1,21 +1,15 @@
...
     entrypoint: mvn
-    args: ['test', '-Pall']
+    args: ['-B', '-e', 'install', '-Pintegration,dsv1_2.12,dsv2', '-s', '/workspace/cloudbuild/gcp-settings.xml', '-Dmaven.repo.local=/workspace/.repository']
     env:
       - 'GOOGLE_CLOUD_PROJECT=${_GOOGLE_CLOUD_PROJECT}'
       - 'TEMPORARY_GCS_BUCKET=${_TEMPORARY_GCS_BUCKET}'
 
-  # 2. Run integration tests
- # - name: 'gcr.io/$PROJECT_ID/dataproc-spark-bigquery-connector-presubmit'
- #   id: 'integration-tests'
```
Comment (himanshukohli09) [inline cloudbuild/cloudbuild.yaml:13]: commented code

**Review by himanshukohli09 — APPROVED** [review]

**Pull request merged**